src/txn/txn_recover.c

```diff
@@ -822,12 +846,164 @@ __recovery_file_scan_prefix(WT_RECOVERY *r, const char *prefix, const char *igno
+    len = strlen("colgroup:") + strlen(format_name) + 1;
+    WT_ERR(__wt_calloc_def(r->session, len, &cgname));
+    WT_ERR(__wt_snprintf(cgname, len, "colgroup:%s", format_name));
+    metac->set_key(metac, cgname);
+    WT_ERR_NOTFOUND_OK(metac->search(metac), true);
+    if (ret == WT_NOTFOUND)
+        has_colgroup = false;
+
+    /* Check file metadata entry exists for uri. */
+    len = strlen("file:") + strlen(format_name) + strlen(".wt") + 1;
+    WT_ERR(__wt_calloc_def(r->session, len, &filename));
+    WT_ERR(__wt_snprintf(filename, len, "file:%s.wt", format_name));
+    metac->set_key(metac, filename);
+    WT_ERR_NOTFOUND_OK(metac->search(metac), true);
+    if (ret == WT_NOTFOUND)
+        has_file = false;
+
+    /* If either colgroup or file metadata entry doesn't exist mark the table for removal. */
+    if (!has_colgroup || !has_file) {
+        WT_ERR(__wt_realloc_def(
+          r->session, &r->remove_uris_allocate, r->nremove_uris + 1, &r->remove_uris));
+        WT_ERR(__wt_strdup(r->session, format_name, &r->remove_uris[r->nremove_uris]));
+        r->nremove_uris++;
+    }
+    ret = 0;
+done:
+err:
+    if (tieredname != NULL)
+        __wt_free(r->session, tieredname);
+    if (cgname != NULL)
+        __wt_free(r->session, cgname);
+    if (filename != NULL)
+        __wt_free(r->session, filename);
+    WT_TRET(metac->close(metac));
+    return (ret);
+}
+
+/*
+ * __remove_files_from_metadata --
+ *     Remove all references from metadata for tables that are marked for removal.
+ */
+static int
+__remove_files_from_metadata(WT_RECOVERY *r)
+{
+    WT_DECL_RET;
+    char *cgname, *tablename, *filename;
+
+    cgname = filename = tablename = NULL;
+
+    for (u_int i = 0; i < r->nremove_uris; i++) {
+        __wt_verbose_level_multi(r->session, WT_VERB_RECOVERY_ALL, WT_VERBOSE_INFO, "%s %s",
+          "removing incomplete table", r->remove_uris[i]);
+
+        /* Remove associated colgroup metadata. */
+        size_t len = strlen("colgroup:") + strlen(r->remove_uris[i]) + 1;
+        WT_ERR(__wt_calloc_def(r->session, len, &cgname));
+        WT_ERR(__wt_snprintf(cgname, len, "colgroup:%s", r->remove_uris[i]));
+        WT_ERR_NOTFOUND_OK(__wt_metadata_remove(r->session, cgname), false);
+
+        /* Remove associated table metadata. */
+        len = strlen("table:") + strlen(r->remove_uris[i]) + 1;
+        WT_ERR(__wt_calloc_def(r->session, len, &tablename));
+        WT_ERR(__wt_snprintf(tablename, len, "table:%s", r->remove_uris[i]));
+        WT_ERR_NOTFOUND_OK(__wt_metadata_remove(r->session, tablename), false);
+
+        /* Remove associated file metadata. */
+        len = strlen("file:") + strlen(r->remove_uris[i]) + strlen(".wt") + 1;
+        WT_ERR(__wt_calloc_def(r->session, len, &filename));
+        WT_ERR(__wt_snprintf(filename, len, "file:%s.wt", r->remove_uris[i]));
+        WT_ERR_NOTFOUND_OK(__wt_metadata_remove(r->session, filename), false);
```

Comment:
Shouldn't we also remove the file on the disk here?

Can we do it by calling `__wt_schema_drop()` since it calls `__wt_meta_track_drop`? I am not sure since we are deferring the remove operations until the meta tracking completes.

Reply:
I have changed it to call `__wt_schema_drop` which will drop the file.